lighthouse-core/test/fraggle-rock/gather/runtime-controller-test.js

```diff
@@ -0,0 +1,174 @@
+ */
+'use strict';
+
+const RuntimeController = require('../../../fraggle-rock/gather/runtime-controller.js');
+const {
+  createMockSendCommandFn: createMockSendCommandFn_,
+  makePromiseInspectable,
+  flushAllTimersAndMicrotasks,
+} = require('../../test-utils.js');
+
+/* eslint-env jest */
+
+jest.useFakeTimers();
+
+// This can be removed when FR becomes the default.
+const createMockSendCommandFn = createMockSendCommandFn_.bind(null, {useSessionId: false});
+
+/** @return {LH.Gatherer.FRProtocolSession} */
+function createMockSession() {
+  const session = /** @type {*} */ ({});
+  session.hasNextProtocolTimeout = jest.fn().mockReturnValue(false);
+  session.setNextProtocolTimeout = jest.fn();
+  return session;
+}
+
+describe('RuntimeController', () => {
+  /** @type {LH.Gatherer.FRProtocolSession} */
+  let sessionMock;
+  /** @type {(controller: RuntimeController, id: number) => Promise<void>} */
+  let forceNewContextId;
+
+  beforeEach(() => {
+    sessionMock = createMockSession();
+
+    forceNewContextId = async (controller, executionContextId) => {
+      controller._session.sendCommand = createMockSendCommandFn()
+        .mockResponse('Page.getResourceTree', {frameTree: {frame: {id: '1337'}}})
+        .mockResponse('Page.createIsolatedWorld', {executionContextId})
+        .mockResponse('Runtime.evaluate', {result: {value: 2}});
+
+      await controller.evaluateAsync('1 + 1', {useIsolation: true});
+      expect(controller.getContextId()).toEqual(executionContextId);
+    };
+  });
+
+  it('should clear context on frame navigations', async () => {
+    const onMock = sessionMock.on = jest.fn();
+
+    const controller = new RuntimeController(sessionMock);
+
+    const frameListener = onMock.mock.calls.find(call => call[0] === 'Page.frameNavigated');
+    expect(frameListener).toBeDefined();
+
+    await forceNewContextId(controller, 42);
+    expect(controller.getContextId()).toEqual(42);
+    frameListener[1]();
+    expect(controller.getContextId()).toEqual(undefined);
+  });
+
+  it('should clear context on execution context destroyed', async () => {
+    const onMock = sessionMock.on = jest.fn();
+
+    const controller = new RuntimeController(sessionMock);
+
+    const executionDestroyed = onMock.mock.calls
+      .find(call => call[0] === 'Runtime.executionContextDestroyed');
+    expect(executionDestroyed).toBeDefined();
+
+    await forceNewContextId(controller, 42);
+    expect(controller.getContextId()).toEqual(42);
+    executionDestroyed[1]({executionContextId: 51});
+    expect(controller.getContextId()).toEqual(42);
+    executionDestroyed[1]({executionContextId: 42});
+    expect(controller.getContextId()).toEqual(undefined);
+  });
+});
+
+describe('.evaluateAsync', () => {
```

Comment:
Should we remove theses tests from `driver-test.js`? Driver will be using the functions from `RuntimeController` directly so I don't see the need to maintain two separate sets of tests.

Reply:
We can, though *something* to test the plumbing is still needed and some of them will stick around.

Reply:
done 👍